mesonbuild/mtest.py

```diff
@@ -1851,36 +1855,51 @@ def test_suitable(self, test: TestSerialisation) -> bool:
 
         return True
 
-    def tests_from_args(self, tests: T.List[TestSerialisation]) -> T.Generator[TestSerialisation, None, None]:
+    @staticmethod
+    def make_selection_checker(patterns: T.List[str]) -> T.Callable[[TestSerialisation], bool]:
         '''
-        Allow specifying test names like "meson test foo1 foo2", where test('foo1', ...)
+        Allow specifying test names like "foo1", where test('foo1', ...)
 
         Also support specifying the subproject to run tests from like
-        "meson test subproj:" (all tests inside subproj) or "meson test subproj:foo1"
-        to run foo1 inside subproj. Coincidentally also "meson test :foo1" to
+        "subproj:" (all tests inside subproj) or "subproj:foo1"
+        to run foo1 inside subproj. Coincidentally also ":foo1" to
         run all tests with that name across all subprojects, which is
-        identical to "meson test foo1"
+        identical to "foo1"
         '''
-        for arg in self.options.args:
-            if ':' in arg:
-                subproj, name = arg.split(':', maxsplit=1)
+        unqualified = set()  # type: T.Set[str]
+        # None represents that any test in the subproject will match.
+        qualified = {}  # type: T.Dict[str, T.Union[T.Set[str], None]]
+
+        for pattern in patterns:
+            if ':' in pattern:
+                subproj, name = pattern.split(':', maxsplit=1)
             else:
-                subproj, name = '', arg
-            for t in tests:
-                if subproj and t.project_name != subproj:
-                    continue
-                if name and t.name != name:
-                    continue
-                yield t
+                subproj, name = '', pattern
+
+            if subproj == '':
+                unqualified.add(name)
+            elif name == '':
+                qualified[subproj] = None
+            else:
+                selected_subproj_tests = qualified.get(subproj, set())
+                if selected_subproj_tests is not None:
+                    selected_subproj_tests.add(name)
+
+        def is_selected(t: TestSerialisation) -> bool:
+            selected_subproj_tests = qualified.get(t.project_name, set())
+
+            return (selected_subproj_tests is None or t.name in selected_subproj_tests or t.name in unqualified)
+
+        return is_selected
 
     def get_tests(self) -> T.List[TestSerialisation]:
         if not self.tests:
             print('No tests defined.')
             return []
 
-        tests = [t for t in self.tests if self.test_suitable(t)]
-        if self.options.args:
-            tests = list(self.tests_from_args(tests))
+        is_selected = TestHarness.make_selection_checker(self.options.args) if self.options.args else lambda _: True
```

Comment:
I don't like the `lambda` here. I think one option could be to add a `default` argument to `make_selection_checker`, that would be `True` for `is_selected`, and `False` for `is_excluded`. If the list of selectors is empty (or None), you simply return the default value.

An other option would be to pass both `args` and `exdlude_tests` to your filter builder, and to do the logic inside it instead of here. Having a single function to filter tests, you could even do something like `tests = list(filter(is_selected, tests))`.